data/registry_model/registry_proxy_model.py

```diff
@@ -0,0 +1,620 @@
+        associated with it.
+
+        Note that there may be multiple records in the same repository for the same blob digest, so
+        the return value of this function may change.
+        """
+        blob = self._get_shared_storage(blob_digest)
+        if blob is None:
+            try:
+                blob = (
+                    ImageStorage.select()
+                    .join(ManifestBlob)
+                    .where(
+                        ManifestBlob.repository_id == repository_ref.id,
+                        ImageStorage.content_checksum == blob_digest,
+                    )
+                    .get()
+                )
+            except ImageStorage.DoesNotExist:
+                return None
+
+        with db_disallow_replica_use():
+            with db_transaction():
+                try:
+                    ImageStoragePlacement.select().where(
+                        ImageStoragePlacement.storage == blob
+                    ).get()
+                except ImageStoragePlacement.DoesNotExist:
+                    try:
+                        self._proxy.blob_exists(blob_digest)
+                    except UpstreamRegistryError:
+                        return None
+                    try:
+                        raw_blob, size = self._proxy.get_blob(blob_digest)
+                    except UpstreamRegistryError:
+                        return None
+                    location_name = storage.preferred_locations[0]
+                    path = get_layer_path(blob)
+                    # TODO: use range requests to download the blob when the server supports it
+                    storage.put_content([location_name], path, raw_blob)
+                    location = get_image_location_for_name(location_name)
+                    if blob.image_size != size:
+                        q = ImageStorage.update(
+                            image_size=size,
+                        ).where(ImageStorage.id == blob.id)
+                        affected = q.execute()
+                        if affected != 1:
+                            return None
+                    ImageStoragePlacement.create(storage=blob, location=location.id)
+
+        return super().get_repo_blob_by_digest(repository_ref, blob_digest, include_placements)
+
+    def convert_manifest(
+        self, manifest, namespace_name, repo_name, tag_name, allowed_mediatypes, storage
+    ):
+        try:
+            parsed = manifest.get_parsed_manifest()
+        except ManifestException:
+            return None
+
+        if parsed.media_type in allowed_mediatypes:
+            return parsed
+
+        schema1 = self.get_schema1_parsed_manifest(
+            manifest, namespace_name, repo_name, tag_name, storage
+        )
+        return schema1.convert_manifest(
+            allowed_mediatypes, namespace_name, repo_name, tag_name, None
+        )
+
+    def get_schema1_parsed_manifest(self, manifest, namespace_name, repo_name, tag_name, storage):
```

Comment:
It doesn't fetch the config, right? It seems it won't provide correct environment variables, etc. after conversion.

Reply:
tracked by https://issues.redhat.com/browse/PROJQUAY-3460